### PR TITLE
Refactor addons into multiple packages

### PR DESCRIPTION
**Pull request by andrewrynhard**
**What this PR does / why we need it**:
https://github.com/kubernetes/kubeadm/issues/348

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
https://github.com/kubernetes/kubeadm/issues/348

### CODE REVIEW

File: cmd/kubeadm/app/cmd/init.go [hunk 1]
```diff
@@ -283,7 +284,11 @@ func (i *Init) Run(out io.Writer) error {
 		return err
 	}
 
-	if err := addonsphase.CreateEssentialAddons(i.cfg, client); err != nil {
+	if err := dnsaddonphase.CreateEsentialDNSAddon(i.cfg, client); err != nil {
```
Comment (andrewrynhard) [inline cmd/kubeadm/app/cmd/init.go]: Wasn't certain if we wanted to keep this as just `addonsphase`. I can change if so.

File: cmd/kubeadm/app/cmd/init.go [hunk 2]
```diff
@@ -283,7 +284,11 @@ func (i *Init) Run(out io.Writer) error {
...
+		return err
+	}
+
+	if err := proxyaddonphase.CreateEssentialProxyAddon(i.cfg, client); err != nil {
```
Comment (andrewrynhard) [inline cmd/kubeadm/app/cmd/init.go]: Wasn't certain if we wanted to keep this as just `addonsphase`. I can change if so.

**Comment by andrewrynhard** [comment]
@luxas I have a few things I can clean up. Still a WIP.

File: cmd/kubeadm/app/phases/addons/proxy/proxy_test.go [hunk 3]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package addons
+package proxy
```
Comment (luxas) [inline cmd/kubeadm/app/phases/addons/proxy/proxy_test.go]: Move over TestCompileManifests with the right test cases.

File: cmd/kubeadm/app/phases/addons/proxy/proxy.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+		ImageOverride:   cfg.UnifiedControlPlaneImage,
+		ClusterCIDR:     getClusterCIDR(cfg.Networking.PodSubnet),
+		MasterTaintKey:  kubeadmconstants.LabelNodeRoleMaster,
+		CloudTaintKey:   algorithm.TaintExternalCloudProvider,
+	})
+	if err != nil {
+		return fmt.Errorf("error when parsing kube-proxy daemonset template: %v", err)
+	}
+
+	err = createKubeProxyAddon(proxyConfigMapBytes, proxyDaemonSetBytes, client)
```
Comment (luxas) [inline cmd/kubeadm/app/phases/addons/proxy/proxy.go]: move into if clause

**Review by luxas — COMMENTED** [review]
I know this is WIP, but left a couple of comments for guidance.

What I yet want to see here is moving the ServiceAccount generation from the `apiconfig` phase to each addon pkg respectively. The same goes for the `kubeadm:node-proxier` ClusterRoleBinding.

Sounds good to you? Thanks!

File: cmd/kubeadm/app/phases/addons/dns/dns.go [hunk 5]
```diff
@@ -81,48 +58,30 @@ func CreateEssentialAddons(cfg *kubeadmapi.MasterConfiguration, client clientset
...
-
-	if err := apiclientutil.CreateConfigMapIfNotExists(client, kubeproxyConfigMap); err != nil {
-		return err
-	}
-
-	kubeproxyDaemonSet := &extensions.DaemonSet{}
-	if err := kuberuntime.DecodeInto(api.Codecs.UniversalDecoder(), daemonSetbytes, kubeproxyDaemonSet); err != nil {
-		return fmt.Errorf("unable to decode kube-proxy daemonset %v", err)
+// CreateServiceAccount creates the necessary serviceaccounts that kubeadm uses/might use, if they don't already exist.
+func CreateServiceAccount(client clientset.Interface) error {
```
Comment (luxas) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: I don't think this needs to be exposed separately.
Well, if we do that it's fine by me, but at least `EnsureDNSAddon` should call it.
(EnsureDNSAddon should not have dependencies)

File: cmd/kubeadm/app/phases/addons/proxy/proxy.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+	apiclientutil "k8s.io/kubernetes/cmd/kubeadm/app/util/apiclient"
+	"k8s.io/kubernetes/pkg/api"
+	k8sversion "k8s.io/kubernetes/pkg/util/version"
+	"k8s.io/kubernetes/plugin/pkg/scheduler/algorithm"
+)
+
+const (
+	// KubeProxyClusterRoleName sets the name for the kube-proxy ClusterRole
+	KubeProxyClusterRoleName = "system:node-proxier"
```
Comment (luxas) [inline cmd/kubeadm/app/phases/addons/proxy/proxy.go]: nit: Add `// TODO: This k8s-generic, well-known constant should be fetchable from an other source, not be in this package`

File: cmd/kubeadm/app/phases/addons/proxy/proxy.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+)
+
+// EnsureProxyAddon creates the kube-proxy and kube-dns addons
+func EnsureProxyAddon(cfg *kubeadmapi.MasterConfiguration, client clientset.Interface) error {
```
Comment (luxas) [inline cmd/kubeadm/app/phases/addons/proxy/proxy.go]: Again, the serviceaccount should be created as the first step here

File: cmd/kubeadm/app/phases/addons/proxy/proxy.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+		if _, err := client.ExtensionsV1beta1().DaemonSets(metav1.NamespaceSystem).Update(kubeproxyDaemonSet); err != nil {
+			return fmt.Errorf("unable to update the kube-proxy daemonset: %v", err)
+		}
+	}
+	return nil
+}
+
+func createClusterRoleBindings(client clientset.Interface) error {
+	// TODO: This ClusterRoleBinding should be created by the kube-proxy phase, not here
```
Comment (luxas) [inline cmd/kubeadm/app/phases/addons/proxy/proxy.go]: can remove this TODO now :)

File: cmd/kubeadm/app/phases/addons/proxy/proxy.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+	return nil
+}
+
+// CreateRBACRules creates the essential RBAC rules for a minimally set-up cluster
+func CreateRBACRules(client clientset.Interface, k8sVersion *k8sversion.Version) error {
+	if err := createClusterRoleBindings(client); err != nil {
+		return err
+	}
+
+	fmt.Println("[proxy] Created RBAC rules")
```
Comment (luxas) [inline cmd/kubeadm/app/phases/addons/proxy/proxy.go]: I don't think we should print this; one line per addon phase is enough

File: cmd/kubeadm/app/phases/addons/proxy/proxy.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+		if !apierrors.IsAlreadyExists(err) {
+			return err
+		}
+	}
+	return nil
+}
+
+// CreateRBACRules creates the essential RBAC rules for a minimally set-up cluster
+func CreateRBACRules(client clientset.Interface, k8sVersion *k8sversion.Version) error {
+	if err := createClusterRoleBindings(client); err != nil {
```
Comment (luxas) [inline cmd/kubeadm/app/phases/addons/proxy/proxy.go]: I really like the idea of having `EnsureFooAddon`, `CreateServiceAccount` and `CreateRBACRules` as the three exported funcs of every addon phase (even if some of them no-op).

File: cmd/kubeadm/app/cmd/init.go [hunk 11]
```diff
@@ -289,16 +290,27 @@ func (i *Init) Run(out io.Writer) error {
...
 	}
 
-	// Create the necessary ServiceAccounts
-	if err := apiconfigphase.CreateServiceAccounts(client); err != nil {
+	if err := apiconfigphase.CreateRBACRules(client, k8sVersion); err != nil {
 		return err
 	}
 
-	if err := apiconfigphase.CreateRBACRules(client, k8sVersion); err != nil {
+	if err := dnsaddonphase.CreateServiceAccount(client); err != nil {
```
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: This should not be necessary

**Review by luxas — COMMENTED** [review]
Thanks a bunch! Only one primary comment; and some small nits to fix:

From golint:
```
W0814 06:27:07.535] cmd/kubeadm/app/phases/addons/dns/manifests.go:20:2: exported const KubeDNSVersion should have comment (or a comment on this block) or be unexported
W0814 06:27:07.535] cmd/kubeadm/app/phases/addons/proxy/manifests.go:20:2: exported const KubeProxyConfigMap should have comment (or a comment on this block) or be unexported
```
From unit/bazel:
```
cmd/kubeadm/app/phases/apiconfig/clusterroles_test.go:60: undefined: CreateServiceAccounts
```

**Review by luxas — APPROVED** [review]
/lgtm

Awesome to get this done!

**Comment by luxas** [comment]
@andrewrynhard kind of annoying, but please fix this 
```
W0814 15:57:13.081] Some packages in hack/.golint_failures do not exist anymore. Please remove them.
W0814 15:57:13.082] 
W0814 15:57:13.082]   cmd/kubeadm/app/phases/addons
```

**Pull request merged**